Develop module first version

ramin-fs opened the pull request:
https://trello.com/c/NJusBZVq/103-pod-checkmodulespodcoverage

Commit:
- Initial commit

Commit:
- Move main module files

Commit:
- Change the basic test file to check new module files instead of old ones
- Change the dist.ini file to use new name

Commit:
- Remove extra files

Commit:
- Remove the extra test file

Commit:
- Change some ignored files in gitignore

Commit:
- Remove all.pm module, maybe we revet is back in the future

Commit:
- Remove travis to prevent tesst run. (we should revert this back in the future)

Commit:
- Update the README.md file with new documentation

Commit:
- Update the README.md file

Commit:
- Replace the actual email address with `email.com` domain

Commit:
- Rename modules name

Commit:
- Change dependencies

Commit:
- Fix basic tests

Commit:
- Add a test file
- Add a simplePackage in test folder

Commit:
- Add Changes file
- Update README.md file
- Complete Syntax tests

Commit:
- Remove debug from coverage

Commit:
- Change linux find with File::Find::Rule

Commit:
- Remove travis file
- Remove podCoverage module
- Remove podSyntax module
- Add PodChange module

Commit:
- Update basic test

Commit:
- Update pods and put some todo

Commit:
- Update test package structure

Commit:
- Change main_module to CoverageChange

Commit:
- Change export model
- Change module usage to use full path

Commit:
- Add nopod test

Commit:
- Update basic test

Commit:
- Complete nopod.t test file

Commit:
- Change Nopod package name to t::Nopod in package file and test also

Commit:
- Make some progress in module tests

Commit:
- Make some changes but maybe we need to revert them

Commit:
- Revert back prev commit

Commit:
- Add ignored_packages to the coverchange module
- Complete correct_pod.t
- Complete nopod.t
- Complete pod_syntax_error.t
- Empty pod_syntax.t and should complete it later

Commit:
- Remove TestPackages folder from tests

Commit:
- Update correct_pod subtest name
- Update pod_syntax_error subtest name

Commit:
- Add a subtest to check module arguments
- Update test variables position

Commit:
- Rename nopod.t to no_pod.t

Commit:
- Update tests

Commit:
- Add ignore package feature as a parameter
- Add ignore package to pod coverage
- Add ignore package to the pod syntax checker

Commit:
- Remove extra line

Commit:
- Add package ignore tests

Commit:
- Update

tom-binary commented on Changes:
```diff
@@ -0,0 +1,7 @@
+{{$NEXT}}
```
This is replaced by the version information

tom-binary commented on Changes:
```diff
@@ -0,0 +1,7 @@
+{{$NEXT}}
+
+Revision history for test-pod-coverage-syntax
```
This should be at the top of the file, *above* the `{{$NEXT}}` template directive

tom-binary commented on Changes:
```diff
@@ -0,0 +1,7 @@
+
+Revision history for test-pod-coverage-syntax
+
+0.01      2020-06-26 00:00:00+00:00 UTC
```
This needs to be removed, it has not been released yet.

tom-binary commented on Changes:
```diff
@@ -0,0 +1,7 @@
+Revision history for test-pod-coverage-syntax
+
+0.01      2020-06-26 00:00:00+00:00 UTC
+        First version, released on an unsuspecting world.
```
```suggestion
        Initial version.
```

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,188 @@
+=head1 DESCRIPTION
+
+Prints the percentage of POD coverage in **TODO** test format for currently existing packages.
+Prints B<not ok> (with respective message) if our latest changes increased/decreased numbers of naked functions in currently existing packages.
```
as mentioned in Trello, should be `C<>`

ramin-fs replied:
I'm going to update all the pods based on the Trello comment. I'll notify you after it is done.

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,188 @@
+
+=head1 DESCRIPTION
+
+Prints the percentage of POD coverage in **TODO** test format for currently existing packages.
```
Don't use markdown in POD - why is this marked as "TODO", what should be here?

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,188 @@
+
+Pod coverage calculator test file.
+
+=head1 SYNOPSIS
```
Synopsis should contain verbatim code showing an example of how to use the module.

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,188 @@
+Prints a proper message for the newly added packages.
+
+prints B<ok> for the files that have no POD syntax error.
+prints B<not ok- There is no POD in the file> if the file has no POD at all. I put this into a TODO test so CircleCI's tests will pass.
```
There is no CircleCI - this is a public module.

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,188 @@
+prints B<not ok- The number of errors in the POD structure> if the file has any error. It causes CircleCI's tests to fail.
+
+=cut
+use strict;
```
POD is paragraph-based, needs blank lines before+after every `=` directive.

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,188 @@
+  FILE_HAS_NO_POD => -1,
+};
+
+our @ISA = qw(Exporter);
```
never use `our @ISA`, it'd be `use parent` and we shouldn't be doing this anyway:

```suggestion
use Exporter qw(import export_to_level);
```

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,188 @@
+
+=over 4
+
+=item * C<directory> An arrayref of directories to check pod against
```
```suggestion
=item * C<directory> - arrayref of directories to check pod against
```

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,188 @@
+    check_pod_coverage($path, $naked_packages, $ignored_packages);
+    check_pod_syntax($path, $ignored_packages);
+
+    Test::Pod::CoverageChange->export_to_level(1, @_);
```
Why is this here?

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,188 @@
+    # Check for newly added packages PODs
+    my @ignored_packages = (keys %$allowed_naked_packages, @$ignored_packages);
+    foreach my $package (Test::Pod::Coverage::all_modules(@$directories)) {
+        next if @ignored_packages && (grep(/^$package$/, @ignored_packages));
```
If this is supposed to support regex and do partial matching, then you wouldn't want the `^$` anchors. If it's a string match, then use a string match...

```suggestion
        next if @ignored_packages && (any { $_ eq $package } @ignored_packages));
```

or preferably:

```
my %ignored;
@ignored{@ignored_packages} = ();
```

so that you can make this a simple `next if exists $ignored{$package}`

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,188 @@
+    }
+
+    my @files_path = File::Find::Rule->file()
+                                     ->name( '*.p[m|l]' )
```
generally alignment should match the level of indentation (1 for the sub, another for the incomplete statement), otherwise you end up with arbitrary spacing based just on function/package names

tom-binary commented on t/correct_pod.t:
```diff
@@ -0,0 +1,21 @@
+use warnings;
```
strict first, then warnings - for some reason all these have them swapped?

ramin-fs replied:
My code editor changes the order. I'll fix it.

tom-binary commented on cpanfile:
```diff
@@ -0,0 +1,27 @@
+requires "Exporter" => "0";
+requires "Moo" => "1.006000";
```
Where are we using this?

tom-binary commented on cpanfile:
```diff
@@ -0,0 +1,27 @@
+requires "Pod::Coverage" => "0";
+requires "Test::More" => "0";
+requires "Test::Pod::Coverage" => "0";
+requires "WebService::Client" => "0.0001";
```
This also seems unused?

tom-binary commented on cpanfile:
```diff
@@ -0,0 +1,27 @@
+requires "Test::More" => "0";
+requires "Test::Pod::Coverage" => "0";
+requires "WebService::Client" => "0.0001";
+requires "base" => "0";
```
not used

tom-binary commented on cpanfile:
```diff
@@ -0,0 +1,27 @@
+requires "WebService::Client" => "0.0001";
+requires "base" => "0";
+requires "constant" => "0";
+requires "strict" => "0";
```
core modules, `strict` and `warnings` are never required in `cpanfile`

Commit:
- Update Changes file
- Reorder strict and warnings
- Remove pod_syntax.t

Commit:
- Remove extra line

Commit:
- Change license string

Commit:
- Update all the PODs
- Fix caller file name

Commit:
- Remove extra lib t imports from test files

Commit:
- Update README file

Commit:
- Update plugin Author
- Update the plugin license

tom-binary commented on cpanfile:
```diff
@@ -0,0 +1,22 @@
+requires "Exporter" => "0";
```
see our other `cpanfile` examples - please don't use `=>` here!

tom-binary commented on cpanfile:
```diff
@@ -0,0 +1,22 @@
+requires "constant" => "0";
+
+on 'test' => sub {
+  requires "File::Spec" => "0";
```
4-space indent

Commit:
- Implement Toms suggestions

Commit:
- Update cpanfile and change the require method
- Update cpanfile indentation

Commit:
- update readme

ramin-fs committed:
Update lib/Test/Pod/CoverageChange.pm

Co-authored-by: Tom Molesworth <tom@binary.com>

ramin-fs committed:
Update lib/Test/Pod/CoverageChange.pm

Co-authored-by: Tom Molesworth <tom@binary.com>

ramin-fs committed:
Update lib/Test/Pod/CoverageChange.pm

Co-authored-by: Tom Molesworth <tom@binary.com>

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,236 @@
+
+=over 4
+
+=item C<$directories> Arrayref|String - I works on all the modules that lies under this given directory(ies).
```
again, please look at other examples - we don't use this style anywhere else in the code

```suggestion
=item C<$directories> - directories to check recursively, supports string or arrayref
```

Commit:
- Update CoverageChange.pm pod
- Update CoverageChange.pm indentation

Commit:
Revert "- update readme"

This reverts commit 5840b98a

Commit:
- Change binary.com to deriv

Commit:
- Import list::util any sub into the CoverageChange package
- Remove extra test lines from correct_pod
- Update some tests

Commit:
- Add done_testing to some subtests
- Add a warning to pod_syntax_error test file
- Remove done_testing parentheses from all tests

Commit:
- Remove use_ok from pods
- Include self library to all the pods
- Update some tests message

ramin-fs committed:
Tom updated an POD item.

Co-authored-by: Tom Molesworth <tom@binary.com>

zakame commented on cpanfile:
```diff
@@ -0,0 +1,22 @@
+    requires "File::Spec";
+    requires "IO::Handle";
+    requires "IPC::Open3";
+    requires "Test::More";
```
Because your tests use `subtest`, this needs 
```suggestion
    requires "Test::More", "0.98";
```

zakame commented on cpanfile:
```diff
@@ -0,0 +1,22 @@
+    requires "IO::Handle";
+    requires "IPC::Open3";
+    requires "Test::More";
+    requires "perl", ">= 5.006";
```
We probably don't need to require such an old perl, and more to the point, probably should go as a `runtime_requires`, not on `test_requires`.

ramin-fs replied:
I completely removed it. I copied it from another module and I didn't know if it better to remove it or keep it. thanks.

zakame commented on dist.ini:
```diff
@@ -0,0 +1,72 @@
+copyright_year   = 2020
+main_module= lib/Test/Pod/CoverageChange.pm
+
+[Git::GatherDir]
```
Maybe https://metacpan.org/pod/Dist::Zilla::PluginBundle::Author::DERIV already provides for the rest of this?

ramin-fs replied:
you mean I replace all the content below this line with `[Author::DERIV]`? If the answer is yes I change it a few minutes ago but I'm not able to install the CPAN module for this plugin. Can you please help me with that?

ramin-fs replied:
Required plugin Dist::Zilla::Plugin::Author::DERIV isn't installed.

Run 'dzil authordeps' to see a list of all required plugins.
You can pipe the list to your CPAN client to install or update them:

    dzil authordeps --missing | cpanm

I noticed that there is a difference between the required module and the module that dzil has tried to install:
// dzil is trying to install this plugin after I use [Author::DERIV]
Dist::Zilla::Plugin::Author::DERIV

// we can install our plugin using this namespace
Dist::Zilla::PluginBundle::Author::DERIV

zakame replied:
I think its supposed to be

```
[@Author::DERIV]
```

precisely because it is a PluginBundle, not just a single Plugin.

See also https://metacpan.org/pod/Dist::Zilla::Starter

Commit:
- Update main module pods

ramin-fs committed:
Update cpanfile

Co-authored-by: Zak B. Elep <zakame@zakame.net>

ramin-fs commented on t/no_pod.t:
```diff
@@ -0,0 +1,39 @@
+subtest 'Module with no pod, unexpected' => sub {
+    test_out("not ok 1 - Pod coverage on $test_module");
+    test_out("not ok 2 # TODO There is no POD in the file $test_module_path.");
+    test_out("#   Failed (TODO) test at $main_module_path line 175.");
```
@zakame can you help me with this, please? I want to get rid of this filename and line number but I've couldn't handle it. Do you have any idea to do that?

zakame replied:
So if I read this right, this TODO fail is raised via https://github.com/binary-com/perl-Test-Pod-CoverageChange/blob/9026e1d4ea3aa1cde400c73b699e3e7bbd1757c8/lib/Test/Pod/CoverageChange.pm#L175-L180 and what you're trying to do is to raise this TODO fail from the context of this test script, not from this `CoverageChange` module, a la Carp.  Per https://metacpan.org/pod/Test::More#Extending-and-Embedding-Test::More, you probably might need to check via the underlying `Test::Builder` instead and use https://metacpan.org/pod/Test::Builder#todo_start and https://metacpan.org/pod/Test::Builder#todo_skip ...

ramin-fs replied:
@zakame I change these parts. Thanks for your great comment.

Commit:
- Remove perl from cpan required modules

Commit:
- Change test more simple function use with Test::Builder functions
- Update failed tests

Commit:
- Change dist.ini content with Author::Deriv

tom-binary commented on cpanfile:
```diff
@@ -0,0 +1,21 @@
+    requires "File::Spec";
+    requires "IO::Handle";
+    requires "IPC::Open3";
+    requires "Test::More", "0.98";
```
```suggestion
    requires "Test::More", ">= 0.98";
```

tom-binary commented on cpanfile:
```diff
@@ -0,0 +1,21 @@
+};
+
+on 'develop' => sub {
+    requires "Test::Pod", "1.41";
```
```suggestion
    requires "Test::Pod", ">= 1.41";
```

tom-binary commented on cpanfile:
```diff
@@ -0,0 +1,21 @@
+    requires "Test::More", "0.98";
+};
+
+on 'configure' => sub {
```
this should be removed entirely

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,246 @@
+package Test::Pod::CoverageChange;
+# ABSTRACT: check perl modules against their pod coverage
+
+=pod
```
not required, should be removed

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,246 @@
+package Test::Pod::CoverageChange;
+# ABSTRACT: check perl modules against their pod coverage
+
```
strict, warnings and POD encoding should always be at the top of the file. Would also need the $VERSION definition.

```
use strict;
use warnings;

our $VERSION = '0.001';
# AUTHORITY

use utf8;

=encoding utf8
```

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,246 @@
+
+    use Test::Pod::CoverageChange qw(check);
+
+    Test::Pod::CoverageChange::check('lib', {
```
if this is being imported, then it wouldn't need to be fully qualified - but I'd suggest renaming, `check` is way too generic if there are other tests in here. Consider `pod_coverage_ok` or similar.

ramin-fs replied:
IMHO `check` functionality can be expressed by the package namespace itself. But if your concern is about the conflict problem then I can change the name.

ramin-fs replied:
BTW your suggested name already in use in this module: https://metacpan.org/pod/Test::Pod::Coverage#pod_coverage_ok(-$module,-%5B$parms,-%5D-$msg-) and we are using it in our module as well.

ramin-fs replied:
I've changed it to `pod_coverage_syntax_ok`

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,246 @@
+Ignores to check every package that we pass as C<@ignored_package>
+Prints a proper message for the newly added packages.
+
+Prints C<ok> for the files that have no POD syntax error.
```
This is inaccurate - this shouldn't be *printing* anything, it's running tests. `prove -v` happens to render this as raw TAP output, but there are many other test harnesses and formatters.

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,246 @@
+Prints a proper message for the newly added packages.
+
+Prints C<ok> for the files that have no POD syntax error.
+Prints C<not ok- There is no POD in the file> if the file has no POD at all. I will consider as a pass test since it is in a todo section.
```
```suggestion
If the file has no POD at all, it will generate a failing TODO test.
```

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,246 @@
+use Pod::Coverage;
+use File::Find::Rule;
+use Test::Pod::Coverage;
+use Module::Path 'module_path';
```
please think about consistency when you're writing code - the line below is using `qw()` for the list, but for some reason you have a different quoting style here...

```suggestion
use Module::Path qw(module_path);
```

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,246 @@
+use Module::Path 'module_path';
+use List::Util qw(any);
+
+our $VERSION = '0.01';
```
Each part of a version in Perl has 3 digits, so I would recommend against 2-digit versions like this. Also, should be at the top of the file.

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,246 @@
+our @EXPORT_OK = qw(check);
+
+# Set caller test file name
+my $caller_test_file = (caller())[1];
```
This should be in the test function, not at the top level.

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,246 @@
+my $caller_test_file = (caller())[1];
+
+# Create a test builder object
+my $Test_Builder = Test::More->builder;
```
This should also be in the test function or at least created on demand.

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,246 @@
+
+=over 4
+
+=item * C<directories> - directories to check recursively, supports string or arrayref
```
inconsistent, doesn't match the naming in the code and misleading because it looks like these are named parameters...

```suggestion
=item * C<$path> - path or arrayref of directories to check (recursively)
```

tom-binary commented on lib/Test/Pod/CoverageChange.pm:
```diff
@@ -0,0 +1,246 @@
+
+=over 4
+
+=item C<directories> - Directories to check recursively, supports string or arrayref
```
same issue here.

tom-binary commented on t/00-basic.t:
```diff
@@ -0,0 +1,11 @@
+use strict;
```
this file can be deleted, it's already covered by dzil

ramin-fs committed:
Update cpanfile

Co-authored-by: Tom Molesworth <tom@binary.com>

ramin-fs committed:
Update cpanfile

Co-authored-by: Tom Molesworth <tom@binary.com>

ramin-fs committed:
Update lib/Test/Pod/CoverageChange.pm

Co-authored-by: Tom Molesworth <tom@binary.com>

ramin-fs committed:
Update lib/Test/Pod/CoverageChange.pm

Co-authored-by: Tom Molesworth <tom@binary.com>

ramin-fs committed:
Update lib/Test/Pod/CoverageChange.pm

Co-authored-by: Tom Molesworth <tom@binary.com>

Commit:
- Update cpanfile
- Update CoverageChange main file

Commit:
- Add file encoding
- Move testbuilder variable to function scope

Commit:
- Update

Commit:
- Update a pod

Commit:
- Merge

Commit:
- Update dist.ini
- Update CoverageChange.pm version and Authority

Commit:
- Update readme file

Commit:
- Update readme file

Commit:
- Add LICENSE
- Add Makefile.PL

Commit:
- Change check sub name to pod_coverage_syntax_check

ramin-fs commented:
Please check [this PR](https://github.com/binary-com/perl-Test-Pod-CoverageChange/pull/2)
